### PR TITLE
Reset iast request context on root span published

### PR DESCRIPTION
**Pull request by manuel-alvarez-alvarez**
# What Does This Do
Resets all IAST request context data structures when the root span of a trace is published.

# Motivation
We observed an akka-http service with a very high number of IAST contexts in the heap probably related to pending traces, once a trace has been published we should not hold down to any references.

# Additional Notes

# Contributor Checklist

- [ ] Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- [ ] Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- [ ] Squash your commits prior merging or merge using GitHub's [Squash and merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits)
- [ ] Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- [ ] Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [APPSEC-55869]

[APPSEC-55869]: https://datadoghq.atlassian.net/browse/APPSEC-55869?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by Mariovido — APPROVED** [review]
LGTM

**Review by jandro996 — APPROVED** [review]

**Review by jandro996 — APPROVED** [review]
LGTM! but there are plenty of failing tests 😅

File: internal-api/src/main/java/datadog/trace/api/iast/IastContext.java [hunk 1]
```diff
@@ -4,11 +4,12 @@
 import datadog.trace.api.gateway.RequestContextSlot;
 import datadog.trace.bootstrap.instrumentation.api.AgentSpan;
 import datadog.trace.bootstrap.instrumentation.api.AgentTracer;
+import java.io.Closeable;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 
 /** Encapsulation for the IAST context, */
-public interface IastContext {
+public interface IastContext extends Closeable {
```
Comment (amarziali) [inline internal-api/src/main/java/datadog/trace/api/iast/IastContext.java:12]: Perhaps implementing `AutoCloseable` is a better choice here if you want to go with try-with-resources

**Comment by manuel-alvarez-alvarez** [comment]
> LGTM! but there are plenty of failing tests 😅

Fixed! :wink:

File: internal-api/src/main/java/datadog/trace/api/iast/IastContext.java (hunk 1 above)
Comment (manuel-alvarez-alvarez) [reply internal-api/src/main/java/datadog/trace/api/iast/IastContext.java:12]: Yep, AFAIK `Closeable` extends `AutoCloseable` so we could eventually use it in a try with resources, but so far it's only used in the span context directly:

```java
    if (this.requestContextDataIast instanceof Closeable) {
      try {
        ((Closeable) this.requestContextDataIast).close();
      } catch (IOException | RuntimeException e) {
        exc = e;
      }
    }
```

**Pull request merged**